### PR TITLE
Add Input Devices Support

### PR DESCRIPTION
**Pull request by emiliopedrollo**
This adds input devices (#1)

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -30,123 +36,187 @@ class Extension {
...
-      this.setupExludedDevicesHandling();
+      this.setupExcludedDevicesHandling();
     });
   }
 
   hideExcludedDevices() {
-    var devices = this._mixer?.getAudioDevicesFromDisplayNames(
-      this._lastExludedDevices!
-    );
+    let devices = this._mixer
```
Comment (marcinjahn) [inline src/extension.ts:55]: `let` is better than `var`, but `const` would be even better :)

File: src/extension.ts [hunk 2]
```diff
@@ -30,123 +36,187 @@ class Extension {
...
+    if (excludedDevices?.includes(deviceName)) {
       delay(200).then(() => {
         // delay due to potential race condition with Quick Setting panel's code
-        this._audioPanel!.removeDevice(deviceId);
+        this._audioPanel!.removeDevice(deviceId, type);
       });
     }
   }
 
   setAllOutputsInSettings() {
```
Comment (marcinjahn) [inline src/extension.ts:189]: This function name needs to be updated

File: src/extension.ts [hunk 3]
```diff
@@ -173,19 +248,27 @@ class Extension {
...
+    this._mixer
+      ?.getAudioDevicesFromDisplayNames(allOutputDevices, "output")
       .filter((n) => n)
-      .map((n) => n!.id);
+      .map((n) => n!.id)
+      .forEach((id) => this._audioPanel!.addDevice(id, "output"));
 
-    devicesToShowIds?.forEach((id) => this._audioPanel!.addDevice(id));
+    this._mixer
+      ?.getAudioDevicesFromDisplayNames(allOutputDevices, "input")
```
Comment (marcinjahn) [inline src/extension.ts:268]: should use `allInputDevices` variable

File: src/settings.ts [hunk 4]
```diff
@@ -26,38 +28,63 @@ export class SettingsUtils {
...
+        : this.getExcludedInputDeviceNames();
 
-    if (currentOutputs.includes(displayName)) {
+    if (currentDevices.includes(displayName)) {
       return;
     }
 
-    const newOutputs = [...currentOutputs, displayName];
-    const settings = this.getSettings();
+    const newOutputs = [...currentDevices, displayName];
```
Comment (marcinjahn) [inline src/settings.ts:53]: should be renamed to something like `newDevices`

File: src/settings.ts [hunk 5]
```diff
@@ -26,38 +28,63 @@ export class SettingsUtils {
...
+    devices.splice(index, 1);
+
+    const setting =
+      deviceType === "output"
+        ? ExcludedOutputNamesSetting
+        : ExcludedInputNamesSetting;
 
     const settings = this.getSettings();
-    settings.set_strv(ExcludedOutputNamesSetting, outputs);
+    settings.set_strv(deviceType, devices);
```
Comment (marcinjahn) [inline src/settings.ts:87]: Should be `settings.set_strv(setting, devices);`

File: src/settings.ts [hunk 6]
```diff
@@ -75,37 +109,51 @@ export class SettingsUtils {
...
+        ? this.getAvailableOutputs()
+        : this.getAvailableInputs();
 
-    if (currentOutputs.includes(displayName)) {
+    if (currentDevices.includes(displayName)) {
       return;
     }
 
-    const newAllOutputs = [...currentOutputs, displayName];
+    const newAllOutputs = [...currentDevices, displayName];
```
Comment (marcinjahn) [inline src/settings.ts:130]: `newAllOutputs` should be renamed to `newAllDevices`.

**Comment by marcinjahn** [comment]
I'd also move the input devices group in `prefs.js` to a new page. That way, outputs/inputs would be in separate tabs of the settings window. I can do that, I'll just wait for you to fix the other issues first.

**Review by marcinjahn — CHANGES_REQUESTED** [review]
Review done, I found just a few small hiccups. Great work!

File: src/extension.ts (hunk 2 above)
Comment (emiliopedrollo) [reply src/extension.ts:189]: Sorry, I don't understand what exactly needs to be updated here

Comment (marcinjahn) [reply src/extension.ts:189]: The function name is still ‘’setAllOutputsInSettings”, although now it handles both outputs and inputs.

Comment (emiliopedrollo) [reply src/extension.ts:189]: Oh. I see it now. I thought the issue was on the line that was changed. 

Unfortunately I won't be able to change it for the next few hours as I'm not at home right now. 

Feel free to rename it on your side if you want to merge it today.

Comment (marcinjahn) [reply src/extension.ts]: Alright :slightly_smiling_face:

**Pull request merged**